File: clojure/Day04.clj
(ns Day04)

; Returns the digits of for a given natural number as a lazy sequence. For example:
; (digits 234789) => (9 8 7 4 3 2)
(defn digits [number]
  (map #(rem % 10) (take-while pos? (iterate #(quot % 10) number))))

; Returns the last digit of the given digits if the digits are in a valid order,
; i.e. never becoming smaller, or false if they do.
(defn is-valid [digits]
  (reduce #(if (> %2 %1) (reduced false) %2) digits))

; Returns a lazy sequence of the sizes of groups in a sequence of digits or nil if there are no groups.
; For example: (9 8 3 3 2 1 1 1) => (2 3)
(defn group-sizes [digits]
  (seq (filter (partial < 1) (vals (frequencies digits)))))

; Given a range of numbers, returns a lazy sequence of digit combinations
; that have both at least one group and are considered valid as defined by is-valid.
(defn possible-passwords [range]
  (filter group-sizes (filter is-valid (map digits range))))

; Parses a range from a raw string, i.e. "123-456" => (range 123 456)
(defn parse-range [raw]
  (apply range (map #(Integer/parseInt %) (.split #"-" raw))))

(def input (parse-range (slurp (first *command-line-args*))))

(let [passwords (possible-passwords input)
      narrowed-passwords (filter #(some (partial = 2) (group-sizes %)) passwords)]
  (println "Amount of possible passwords:" (count passwords))
  (println "Amount with extra condition:" (count narrowed-passwords)))

File: haskell/Day04.hs
module Main where

import Common

-- Validate an integer by pattern (1)
validate :: Int -> Bool
validate number = size && increase number 100000 False
  where
    size = number `div` 100000 >= 1 && number `div` 100000 <= 10
    increase a pos b
      | pos == 1  = b
      | otherwise = (a `div` pos) <= (a - (a `div` pos) * pos) `div` (pos `div` 10) && increase (a - (a `div` pos) * pos) (pos `div` 10) bool'
    
      where bool' = if b then b else ((a `div` pos) == (a - (a `div` pos) * pos) `div` (pos `div` 10))

-- Validate an integer by pattern (2)
validate2 :: Int -> Bool
validate2 number = validate number && elem True (map (containsTwo digits') digits')
  where digits' = digits number 6

-- Run the validation function between a given range
run :: Int -> Int -> (Int -> Bool) -> Int
run from to func = length (filter func [from..to])

-- Get the list of digits from a number a
digits :: Int -> Int -> [Int]
digits a 0 = []
digits a length = digit : digits (a - (digit * 10^ (length - 1))) (length - 1)
  where digit = a `div` (10 ^ (length - 1))

-- Check if a list contains 2 times the same element
containsTwo :: (Eq t) => [t] -> t -> Bool
containsTwo list t = (helper list 0) == 2
  where
    helper [] i     = i
    helper (x:xs) i = if x == t then helper xs (i + 1) else helper xs i
    
{- Handle input and solve executions -}

-- Solve first challenge
solve1 :: Input -> Int
solve1 (Line (Just line)) = run (read (splitted !! 0)) (read (splitted !! 1)) validate
  where splitted = split line '-'

-- Solve second challenge
solve2 :: Input -> Int
solve2 (Line (Just line)) = run (read (splitted !! 0)) (read (splitted !! 1)) validate2
  where splitted = split line '-'

-- Print challenge result
main = do
  solve "Amount of different passwords"     (Line Nothing) (solve1)
  solve "Amount of different passwords (2)" (Line Nothing) (solve2)